### ISSUE TITLE
Restore edit mode after command

### ISSUE THREAD

**Issue by agoose77** [jupyterlab/jupyterlab#10509]
I prefer to use keyboard shortcuts in most applications, and the workflow in JupyterLab is great. I do, however, find using commands in edit mode to be slightly long-winded; e.g. to change a cell to Markdown whilst in edit mode requires
* <kbd>Esc</kbd>
* <kbd>M</kbd>
* <kbd>Enter</kbd>

I wonder how difficult it would be to use <kbd>Esc</kbd> as a modifier, whereby the Edit mode is restored upon release, e.g. the previous sequence becomes 
* <kbd>Esc</kbd> + <kbd>M</kbd>
(with hidden timing context)

I don't know how feasible this is, i.e. whether <kbd>Esc</kbd> is treated differently to other modifiers, or whether we want to reserve the modifier for other commands. 

This is something I'd like to take a look at at some point, but as I don't have a lot of time right now I'm opening this issue in case someone else gets there first / for future reference!

**Comment by jasongrout** [comment]
When you say Esc + M, do you mean you are holding down escape and M at the same time?

The keyboard shortcut system does support shortcuts that are sequential key presses. However, that will introduce a delay in the action in the usual case, since after pressing escape, it needs to wait for the timeout to see there are no other keys before the action can occur.

Escape is not treated any different in the shortcut system compared to any other key.

You can experiment with this just by customizing your shortcuts in the advanced settings. Note, for example, the escape key is:

```
        {
            "command": "notebook:enter-command-mode",
            "keys": [
                "Escape"
            ],
            "selector": ".jp-Notebook.jp-mod-editMode"
        },
```
In order to execute several commands in sequence in a shortcut, you can use this new `runAllEnabled` command in the next beta of JLab 3.1: https://github.com/jupyterlab/jupyterlab/pull/10215

**Comment by agoose77** [comment]
> When you say Esc + M, do you mean you are holding down escape and M at the same time?

I was thinking of a release modifier, which I don't think we currently support. The idea would be that the <kbd>Esc</kbd> key enters command mode on hold, and leaves on exit. To integrate this with the existing behaviour, this new feature would only apply if any additional keys are depressed whilst <kbd>Esc</kbd> is held.

**Comment by jasongrout** [comment]
We don't treat escape as a modifier, i.e., there is no way in the shortcut system right now to ask for a shortcut that is esc and m being held down at the same time. The shortcut system [closely follows](https://github.com/jupyterlab/lumino/blob/b71e3ba2ceb0a37d25cde02022f72154e1ff2d10/packages/commands/src/index.ts#L898-L901) the browser capabilities, and only uses a subset of the [browser-recognized modifier keys](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState).

However, the shortcut system could in theory keep track of keydown and keyup events for user-selectable modifier keys. This would involve a PR to the Lumino [commands package](https://github.com/jupyterlab/lumino/blob/b71e3ba2ceb0a37d25cde02022f72154e1ff2d10/packages/commands/src/index.ts), which is providing the keyboard shortcut system. It would have to pay attention to corner cases like focus leaving the application while a user modifier was down (thereby potentially losing the keyup), etc.

**Comment by jasongrout** [comment]
What maybe could work right now is pressing shift-escape (or modifier-escape) to jump to edit mode for one command. For example, we could define ["Accel Escape", "M"] for changing the current cell to markdown and jumping back to edit mode. 

Or I think if there were popular commands that need to be done in edit mode, we could just define edit mode shortcuts directly. For example, we could just define ["Accel M"] in edit mode to change the current cell to markdown and stay in edit mode.

**Comment by jasongrout** [comment]
I too find myself often a bit frustrated with inserting a new cell, starting to type, realizing it is the wrong type, then spending a second or two switching the type with several keystrokes. For this particular case, we could do what many other notebook platforms do and make inserting a markdown cell a separate command, e.g., A/B insert a code cell, while Accel-A/B insert a markdown cell, for example.

**Comment by yashpalgoyal1304** [comment]
i agree with basic premise of this issue, that is the underlying problem. 

But i would be against making the purposed solution of using escape to be used as a modifier key.
* Escape is normally there to function independently and immediately. 
* And it would be extremely unconventional behaviour to have it as a modifier key. 
* Besides, escape key is on the far far top left of the keyboard. 
* There already exists `^m` (i.e. <kbd>Ctrl</kbd> + <kbd>m</kbd>) for switching back to command mode.

I dont have any alternate solutions to propose for. The thoughts of @ jasongrout seem to be on the right track.
